keygen/src/xcr_comm.c

```diff
@@ -0,0 +1,149 @@
+#include <string.h>
+#include "xcr_comm.h"
+#include "json_model.h"
+
+static const char *XAP_XCR_URL = "https://xaptum.io/api/xcr/v2/";
+static const char *XCR_FUNC_AUTH = "xauth";
+
+struct MemoryStruct {
+    char *memory;
+    size_t size;
+};
+
+
+/**
+ * xcr_curl_cb() - Set common CURL parameters
+ * @contents: A data buffer to send to the remote host
+ * @size: Always 1 (according to CURL docs)
+ * @nmemb: The size of @contents
+ * @userp: The user provided memory buffer to write into
+ *
+ * This is the callback function provided in the CURL docs to copy data from
+ * its internal receive buffer to a user defined one.
+ *
+ * Return: The total number of bytes copied
+ */
+static size_t
+xcr_curl_cb(void *contents, size_t size, size_t nmemb, void *userp)
+{
+    size_t realsize = size * nmemb;
+    struct MemoryStruct *mem = (struct MemoryStruct *)userp;
+    printf("Calling realloc for size = %d+%d+1=%d\n",mem->size, realsize, mem->size + realsize + 1);
+    char *ptr = realloc(mem->memory, mem->size + realsize + 1);
+    if(!ptr) {
+        /* out of memory! */
+        printf("not enough memory (realloc returned NULL)\n");
+        return 0;
+    }
+
+    mem->memory = ptr;
+    memcpy(&(mem->memory[mem->size]), contents, realsize);
+    mem->size += realsize;
+    mem->memory[mem->size] = 0;
+
+    return realsize;
+}
+
+/**
+ * set_std_curl_opts() - Set common CURL parameters
+ * @curl: A pointer to the curl handler
+ * @headers: A pointer to the curl_slist containing the headers
+ * @url: A string representing the URL
+ * @data: A buffer containing al the data to send
+ * @target: The buffer to return to
+ *
+ * This is a helper function that sets common parameters for xending to XMB
+ *
+ * Return: None
+ */
+static void set_std_curl_opts(CURL *curl, struct curl_slist *headers, const char *url, const char *data, struct MemoryStruct *target)
+{
+    headers = curl_slist_append(headers, "Accept: application/json");
+    headers = curl_slist_append(headers, "Content-Type: application/json");
+
+    curl_easy_setopt(curl, CURLOPT_URL, url);
+    curl_easy_setopt(curl, CURLOPT_POST, 1L);
+    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+    curl_easy_setopt(curl, CURLOPT_USERAGENT, "libcurl-agent/1.0");
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, xcr_curl_cb);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)target);
+    curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+}
+
+/**
+ * xcr_send_login() - Send an Auth request to XCR
+ * @session: A pointer to the session to fill
+ * @credentials: A pointer to the credentials to use to log in
+ *
+ * Log in to XCR, populate a session structure for use in subsequent calls
+ * and return the fields sent back from the API
+ *
+ * Return: The data send back from the server
+ */
+struct xcr_auth *xcr_send_login(struct xcr_session *session, struct xcr_login *credentials)
+{
+    json_t *xcr_proc_auth_resp = NULL;
```

Comment:
Wow, the cURL API is messy!  Lots of code for what would be one line in Python or another memory-managed, high-level language!

I expect that you'll be able to refactor a bunch of this into smaller utility methods, to make long term maintenance easier.  But probably best to wait until after/when adding the IAM API, to really see what the commonalities are.